classes/page_group.php

```diff
@@ -0,0 +1,114 @@
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with Moodle.  If not, see <http://www.gnu.org/licenses/>.
+
+namespace tool_excimer;
+
+use core\persistent;
+
+/**
+ * Metadata for profile groups.
+ *
+ * @package   tool_excimer
+ * @author    Jason den Dulk <jasondendulk@catalyst-au.net>
+ * @copyright 2022, Catalyst IT
+ * @license   http://www.gnu.org/copyleft/gpl.html GNU GPL v3 or later
+ */
+class page_group extends persistent {
+
+    /** The name of the database table. */
+    public const TABLE = 'tool_excimer_page_groups';
+
+    /**
+     * Returns the fuzzy counts for duration.
+     *
+     * @return array
+     */
+    protected function get_fuzzydurationcounts(): array {
+        $raw = $this->raw_get('fuzzydurationcounts');
+        if (empty($raw)) {
+            return [];
+        }
+
+        return json_decode($raw, true);
+    }
+
+    /**
+     * Sets the fuzzy duration counts.
+     *
+     * @param array $data
+     */
+    protected function set_fuzzydurationcounts(array $data) {
+        $this->raw_set('fuzzydurationcounts', json_encode($data));
+    }
+
+    /**
+     * Records fuzzy count metadata about a page group.
+     *
+     * @param profile $profile The profile to pull the information from.
+     */
+    public static function record_fuzzy_counts(profile $profile) {
+
+        // Get the profile group record, creating a new one if one does not yet exist.
+        $month = userdate(time() - 360, '%Y%m'); // YYYYMM format.
+        $profilegroup = self::get_record(['name' => $profile->get('groupby'), 'month' => $month]);
+        if ($profilegroup === false) {
+            $profilegroup = new page_group(0, (object) ['name' => $profile->get('groupby'), 'month' => $month]);
+        }
+
+        $existing = $profilegroup->to_record();
+
+        // Fuzzy increment the count.
+        $current = $profilegroup->get('fuzzycount');
+        $new = manager::approximate_increment($current);
+        $profilegroup->set('fuzzycount', $new);
+
+        // Fuzzy increment the duration count.
+        $duration = $profile->get('duration');
+        $fuzzydurationcounts = $profilegroup->get('fuzzydurationcounts');
+        $fuzzydurationsum = $profilegroup->get('fuzzydurationsum');
+        $high = 1;
+        while ($duration >= $high) {
+            $high *= 2;
+        }
+        $low = ($high == 1) ? 0 : $high / 2;
+        $median = ($high + $low) / 2.0;
+        $fuzzydurationsum += $median;
+
+        $fuzzydurationcounts[$high] = manager::approximate_increment($fuzzydurationcounts[$high] ?? 0);
+
+        $profilegroup->set('fuzzydurationcounts', $fuzzydurationcounts);
+        $profilegroup->set('fuzzydurationsum', $fuzzydurationsum);
+
+        if ($existing != $profilegroup->to_record()) {
+            $profilegroup->save();
```

Comment:
I don't like display logic mutating the data, eventually we want to push this off to the data ware house so the data needs to be ready to go without a human having viewed this report. so either calculate it when it changes which should be not too bad, or worst case in a task but I would like to avoid that

Reply:
sorry I've misunderstood this code, this looks good to me